connectors/amazons3/connector/src/main/java/org/apache/manifoldcf/crawler/connectors/amazons3/AmazonS3Connector.java

```diff
@@ -0,0 +1,917 @@
+    }
+    Logging.connectors.warn("Amazons3: IO exception: " + e.getMessage(), e);
+    long currentTime = System.currentTimeMillis();
+    throw new ServiceInterruption("IO exception: " + e.getMessage(), e,
+        currentTime + 300000L, currentTime + 3 * 60 * 60000L, -1, false);
+  }
+
+  private static void handleResponseException(ResponseException e)
+      throws ManifoldCFException, ServiceInterruption {
+    throw new ManifoldCFException("Unexpected response: " + e.getMessage(),
+        e);
+  }
+
+  @Override
+  public void processDocuments(String[] documentIdentifiers,
+      IExistingVersions statuses, Specification spec,
+      IProcessActivity activities, int jobMode,
+      boolean usesDefaultAuthority) throws ManifoldCFException,
+      ServiceInterruption {
+    AmazonS3 amazons3Client = getClient();    
+    documentProcess.doProcessDocument(documentIdentifiers, statuses, spec,
+        activities, jobMode, usesDefaultAuthority, amazons3Client);
+
+  }
+
+  protected static class GetSeedsThread extends Thread {
+    protected Throwable exception = null;
+
+    protected String[] bucketsToBeRemoved;
+
+    protected AmazonS3 s3client = null;
+
+    protected XThreadBuffer<S3Artifact> seedBuffer;
+
+    public XThreadBuffer<S3Artifact> getBuffer() {
+      return seedBuffer;
+    }
+
+    public void setBuffer(XThreadBuffer<S3Artifact> buffer) {
+      this.seedBuffer = buffer;
+    }
+
+    public GetSeedsThread(AmazonS3 s3, String[] buckets) {
+      super();
+      this.bucketsToBeRemoved = buckets;
+      this.s3client = s3;
+      seedBuffer = new XThreadBuffer<S3Artifact>();
+      setDaemon(true);
+    }
+
+    @Override
+    public void run() {
+      try {
+        // push the keys for all documents
+        processSeeds();
+      }
+      catch (Exception e) {
+
+        Logging.connectors.error(e);
+        this.exception = e;
+      }
+      finally {
+        seedBuffer.signalDone();
+      }
+    }
+
+    private void processSeeds() {
+
+      if (s3client != null) {
+
+        List<Bucket> listBuckets = s3client.listBuckets();
+        List<String> refinedBuckets = new ArrayList<String>();
+        
+        for (Bucket bucket : listBuckets) {
+          if (bucketsToBeRemoved != null && bucketsToBeRemoved.length > 0 &&  !Arrays.asList(bucketsToBeRemoved).contains(
+              bucket.getName())) {
+            refinedBuckets.add(bucket.getName());
+          }
+          else{
+            refinedBuckets.add(bucket.getName());
+          }
+        }
+
+        for (String bucket : refinedBuckets) {
+          String bucketName = bucket;
+          try {
+            PushSeeds(bucketName);
+          }
+          catch (Exception e) {
+            Logging.connectors.error(e);
+          }
+        }
+
+      }
+      else {
+        Logging.connectors.info("Could not connect amazon");
+      }
+    }
+
+    private void PushSeeds(String bucketName) {
+      try {
+        ObjectListing objectListing = s3client
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

This code uses an outdated API. [ListObjectsV2 is the revised List Objects API, and we recommend you use this revised API for new application developments](https://docs.aws.amazon.com/AmazonS3/latest/API/v2-RESTBucketGET.html).